### PR TITLE
FF-2945 feat: Python SDK

### PR DESCRIPTION
**Pull request by rasendubi**
This PR is ready for review.

It's probably best to start review with the announcement to understand main changes: https://github.com/Eppo-exp/eppo-docs/pull/454

### CODE REVIEW

File: python-sdk/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+requires = ["maturin>=1.7,<2.0"]
+build-backend = "maturin"
+
+[project]
+name = "eppo-server-sdk"
+description = "Eppo SDK for Python"
+readme = "README.md"
+authors = [{ name = "Eppo", email = "eppo-team@geteppo.com" }]
+license = { file = "LICENSE" }
+requires-python = ">=3.8"
```
Comment (rasendubi) [inline python-sdk/pyproject.toml:11]: ☝️ bumping required python version from 3.6 to 3.8

File: python-sdk/pytest.ini [hunk 2]
```diff
@@ -0,0 +1,3 @@
+[pytest]
+markers =
+    rust_only: mark a test as only passing on Python-on-Rust SDK.
```
Comment (rasendubi) [inline python-sdk/pytest.ini:3]: ☝️ All tests without this marker pass on both python-on-rust and native python implementations

File: .github/workflows/python.yml [hunk 3]
```diff
@@ -0,0 +1,280 @@
+# This file is autogenerated by maturin v1.7.0
```
Comment (aarsilv) [inline .github/workflows/python.yml:1]: 👀  `maturin` looks pretty sweet!

File: eppo_core/src/attributes.rs [hunk 4]
```diff
@@ -46,3 +46,34 @@ impl From<&str> for AttributeValue {
...
+    use pyo3::{exceptions::PyTypeError, prelude::*, types::*};
+
+    use super::*;
+
+    impl<'py> FromPyObject<'py> for AttributeValue {
+        fn extract_bound(value: &Bound<'py, PyAny>) -> PyResult<AttributeValue> {
+            if let Ok(s) = value.downcast::<PyString>() {
+                return Ok(AttributeValue::String(s.extract()?));
+            }
+            // In Python, Bool inherits from Int, so it must be checked first here.
```
Comment (aarsilv) [inline eppo_core/src/attributes.rs:61]: Appreciate the helpful comments

File: eppo_core/src/poller_thread.rs [hunk 5]
```diff
@@ -28,7 +28,7 @@ pub struct PollerThreadConfig {
 
 impl PollerThreadConfig {
     /// Default value for [`PollerThreadConfig::interval`].
-    pub const DEFAULT_POLL_INTERVAL: Duration = Duration::from_secs(5 * 60);
+    pub const DEFAULT_POLL_INTERVAL: Duration = Duration::from_secs(30);
```
Comment (aarsilv) [inline eppo_core/src/poller_thread.rs:31]: 👍  Thanks for conforming to our SDK standard of default 30 second polling

File: mock-server/prepare.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+// This is a prepare script that is automatically run before "start" command.
```
Comment (aarsilv) [inline mock-server/prepare.js:1]: Why the change to JS and requirement of node?

File: python-sdk/.gitignore [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+*.log
+*.pot
+
+.DS_Store
+
+# Sphinx documentation
+docs/_build/
+
+# PyCharm
+.idea/
```
Comment (aarsilv) [inline python-sdk/.gitignore:66]: 💪

File: python-sdk/README.md [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+
+[Eppo](https://www.geteppo.com/) is a modular flagging and experimentation analysis tool. Eppo's Python SDK is built to make assignments in multi-user server side contexts. Before proceeding you'll need an Eppo account.
+
+## Features
+
+- Feature gates
+- Kill switches
+- Progressive rollouts
+- A/B/n experiments
+- Mutually exclusive experiments (Layers)
```
Comment (aarsilv) [inline python-sdk/README.md:13]: while in here, how about adding:
```md
- Holdouts
- Contextual multi-armed bandits
```

File: python-sdk/README.md [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+
+## Initialization options
+
+The `init` function accepts the following optional configuration arguments.
+
+| Option | Type | Description | Default |
+| ------ | ----- | ----- | ----- |
+| **`assignment_logger`**  | AssignmentLogger | A callback that sends each assignment to your data warehouse. Required only for experiment analysis. See [example](#assignment-logger) below. | `None` |
+| **`is_graceful_mode`** | bool | When true, gracefully handles all exceptions within the assignment function and returns the default value. | `True` |
+| **`poll_interval_seconds`** | int | The interval in seconds at which the SDK polls for configuration updates. | `300` |
```
Comment (aarsilv) [inline python-sdk/README.md:84]: default is now `30` 🏎️

File: python-sdk/README.md [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+| **`poll_jitter_seconds`** | int | The jitter in seconds to add to the poll interval. | `30` |
```
Comment (aarsilv) [inline python-sdk/README.md:85]: does rust do this? Do we pass this through? I wonder if we just drop this. In other SDKs we just do hard-coded 10% polling interval

File: python-sdk/pyproject.toml (hunk 1 above)
Comment (aarsilv) [reply python-sdk/pyproject.toml:11]: Sounds good, given 3.6 hit end of life two years ago 😬

File: python-sdk/python/eppo_client/__init__.py [hunk 11]
```diff
@@ -0,0 +1,29 @@
...
+# to make submodules re-exportable (by tweaking sys.modules) but it
+# has some drawbacks.
+#
+# It's more straightforward to add a bit of Python to re-export items
+# at different locations.
+#
+# This __init__.py just re-exports everything from the Rust module.[2]
+#
+# [1]: https://github.com/PyO3/pyo3/issues/759
+# [2]: https://www.maturin.rs/project_layout#pure-rust-project
```
Comment (aarsilv) [inline python-sdk/python/eppo_client/__init__.py:14]: 👍

File: python-sdk/python/eppo_client/_eppo_client.pyi [hunk 12]
```diff
@@ -0,0 +1,171 @@
...
+        default: bool,
+    ) -> bool: ...
+    def get_json_assignment(
+        self,
+        flag_key: str,
+        subject_key: str,
+        subject_attributes: Dict[str, Union[str, int, float, bool, None]],
+        default: Any,
+    ) -> Any: ...
+    def get_string_assignment_details(
```
Comment (aarsilv) [inline python-sdk/python/eppo_client/_eppo_client.pyi:75]: `details` 📈

File: python-sdk/python/eppo_client/_eppo_client.pyi [hunk 13]
```diff
@@ -0,0 +1,171 @@
...
+        subject_context: (
+            ContextAttributes | Dict[str, Union[str, int, float, bool, None]]
+        ),
+        actions: (
+            Dict[str, ContextAttributes]
+            | Dict[str, Dict[str, Union[str, int, float, bool, None]]]
+        ),
+        default: str,
+    ) -> EvaluationResult: ...
+    def get_bandit_action_details(
```
Comment (aarsilv) [inline python-sdk/python/eppo_client/_eppo_client.pyi:123]: 🤯

File: python-sdk/python/eppo_client/_eppo_client.pyi [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+    def get_bandit_action(
+        self,
+        flag_key: str,
+        subject_key: str,
+        subject_context: (
+            ContextAttributes | Dict[str, Union[str, int, float, bool, None]]
+        ),
+        actions: (
+            Dict[str, ContextAttributes]
+            | Dict[str, Dict[str, Union[str, int, float, bool, None]]]
```
Comment (aarsilv) [inline python-sdk/python/eppo_client/_eppo_client.pyi:119]: For later, we should also support contextless actions. I know we try to steer customers away from this, but I've seen this use case come up on many calls.

File: python-sdk/src/client.rs [hunk 15]
```diff
@@ -0,0 +1,709 @@
...
+    }
+
+    fn get_string_assignment_details(
+        slf: &Bound<EppoClient>,
+        flag_key: &str,
+        subject_key: &str,
+        subject_attributes: Attributes,
+        default: Py<PyString>,
+    ) -> PyResult<EvaluationResult> {
+        slf.get().get_assignment_details(
```
Comment (aarsilv) [inline python-sdk/src/client.rs:239]: would it make sense for the underlying rust function to have details passed in as boolean like pass in type? Then one less "public" function upstream SDKs need to worry about.

File: python-sdk/src/configuration.rs [hunk 16]
```diff
@@ -0,0 +1,70 @@
...
+        PySet::new_bound(py, self.configuration.flags.flags.keys())
+    }
+
+    // Returns a set of all bandit keys that have been initialized.
+    // This can be useful to debug the initialization process.
+    fn get_bandit_keys<'py>(&'py self, py: Python<'py>) -> PyResult<Bound<PySet>> {
+        PySet::new_bound(
+            py,
+            self.configuration
+                .bandits
```
Comment (aarsilv) [inline python-sdk/src/configuration.rs:46]: minor but we're deprecating the array-valued `bandits` in favor of object-valued `banditReferences`. Both have a top-level object though whose keys are the bandit keys, so should be an easy swap.

File: python-sdk/tests/test_bandit_result.py [hunk 17]
```diff
@@ -0,0 +1,22 @@
...
+    result = BanditResult(variation="variation", action=None)
+    assert result.to_string() == "variation"
+
+
+def test_bandit_result_to_string_action():
+    result = BanditResult(variation="variation", action="action")
+    assert result.to_string() == "action"
+
+
+# Native python implementation does not convert variation to string.
```
Comment (aarsilv) [inline python-sdk/tests/test_bandit_result.py:18]: Interesting.

But yeah in the Eppo application UI, we only allow creating string-valued bandit flags

**Review by aarsilv — COMMENTED** [review]
💪 💪 💪  Awesome stuff! 📈 

This is really exciting. I'm not approving because I don't know Python well and Rust at all. Although this makes me want to learn the latter 🤓 

Thanks to helpful comments and method/variable names, I think I was able to decently follow. 🙏 Left a bunch of minor comments you can take or leave as you see fit!

File: mock-server/prepare.js (hunk 6 above)
Comment (rasendubi) [reply mock-server/prepare.js:1]: This is to make it cross-platform and easier to run in CI. Alpine linux doesn't have `bash` by default (it has `sh`) and Windows doesn't have any posix shell by default

File: python-sdk/README.md (hunk 10 above)
Comment (rasendubi) [reply python-sdk/README.md:85]: Yes. This configuration is passed through and the core applies the jitter

File: python-sdk/python/eppo_client/_eppo_client.pyi (hunk 14 above)
Comment (rasendubi) [reply python-sdk/python/eppo_client/_eppo_client.pyi:119]: Do you mean supporting `actions: Dict[str, None]` or `list[str]`? We should be able to add this in the future in a backward-compatible way

File: python-sdk/src/client.rs (hunk 15 above)
Comment (rasendubi) [reply python-sdk/src/client.rs:239]: Do you mean a type-level parameter like this?
```rs
get_assignment::<WithDetails>(...)
```
It's possible but is a bit convoluted to implement (requires setting up traits and proxying implementation through trait), so isn't worth it for internal function

File: python-sdk/tests/test_bandit_result.py (hunk 17 above)
Comment (rasendubi) [reply python-sdk/tests/test_bandit_result.py:18]: 4.0 extends `BanditResult` to `EvaluationResult` so different-typed variations are allowed. `BanditResult` is now an alias to new `EvaluationResult`

Something like this:
```python
# old BanditResult:
class BanditResult:
    variation: str
    action: str | None

# new EvaluationResult:
class EvaluationResult:
    variation: Any
    action: str | None
    evaluation_details: Any | None
```

File: python-sdk/README.md (hunk 8 above)
Comment (rasendubi) [reply python-sdk/README.md:13]: added

File: python-sdk/README.md (hunk 9 above)
Comment (rasendubi) [reply python-sdk/README.md]: good catch! fixed

File: python-sdk/.gitignore (hunk 7 above)
Comment (rasendubi) [reply python-sdk/.gitignore:66]: hehe, I don't actually use intellij products. this was auto-generated

**Comment by leoromanovsky** [comment]
I'm able to install with `pip` but not able to integrate it into a sample program:

```
➜  python-sdk git:(feat-python-sdk) ✗ pip install -e .
Obtaining file:///Users/leo/src/rust-sdk/python-sdk
  Installing build dependencies ... done
  Checking if build backend supports build_editable ... done
  Getting requirements to build editable ... done
  Preparing editable metadata (pyproject.toml) ... done
Building wheels for collected packages: eppo-server-sdk
  Building editable for eppo-server-sdk (pyproject.toml) ... done
  Created wheel for eppo-server-sdk: filename=eppo_server_sdk-4.0.0-cp311-cp311-macosx_11_0_arm64.whl size=4297 sha256=d3076a814895414aff8a1e5b29785242d953393c21c320d91d93054dc412e1cb
  Stored in directory: /private/var/folders/1f/95qsl9p95fz0k6jgh5cgcd3r0000gn/T/pip-ephem-wheel-cache-wsivpflm/wheels/05/a2/8b/febb161210fb0561c324e64b7ff9400edef701dbe616d25850
Successfully built eppo-server-sdk
Installing collected packages: eppo-server-sdk
  Attempting uninstall: eppo-server-sdk
    Found existing installation: eppo-server-sdk 4.0.0
    Uninstalling eppo-server-sdk-4.0.0:
      Successfully uninstalled eppo-server-sdk-4.0.0
Successfully installed eppo-server-sdk-4.0.0

[notice] A new release of pip is available: 24.1.2 -> 24.2
[notice] To update, run: python3.11 -m pip install --upgrade pip
```

❌ 

```
➜  python-sdk git:(feat-python-sdk) ✗ python3 -c "from eppo_client import EppoClient; print(EppoClient)"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'eppo_client'
```

The existing sample program I have is:

```
import eppo_client
import time
from eppo_client.config import Config
from eppo_client.assignment_logger import AssignmentLogger

# Set this as your Eppo API Key - https://docs.geteppo.com/prerequisites/feature-flagging/randomization-sdk/api-keys
api_key = ""

# Set this one of your Eppo experiment keys
experiment_key = ""

class ExampleAssignmentLogger(AssignmentLogger):
  def log_assignment(self, assignment_event):
    return print("Assigned experiment %s variation %s to subject %s" %(assignment_event["experiment"], assignment_event["variation"], assignment_event["subject"]))

if __name__ == "__main__":
  if not api_key:
    print("Edit example.py to enter your Eppo API key")
    exit(1)
  
  config = Config(api_key=api_key, assignment_logger=ExampleAssignmentLogger())
  client = eppo_client.init(config)
  time.sleep(1) # wait for initialization
  variation = client.get_assignment("my-user", experiment_key)
  if variation is None:
    print("No variation assigned. Check that the experiment is running.")
```

Are adjustments needed to it or to `pyproject.toml` ?

**Comment by rasendubi** [comment]
@leoromanovsky that's interesting. The import should work without any changes.

Could you give me the output of the following:
- `which pip`
- `which python3`
- `ls python/eppo_client`

also, does it work with non-editable install (`pip install .`)?

**Comment by leoromanovsky** [comment]
> * ls python/eppo_client

```
➜  python-sdk git:(feat-python-sdk) ✗ pwd
/Users/leo/src/rust-sdk/python-sdk
➜  python-sdk git:(feat-python-sdk) ✗ which pip
/opt/homebrew/bin/pip
➜  python-sdk git:(feat-python-sdk) ✗ which pip3
/opt/homebrew/bin/pip3
➜  python-sdk git:(feat-python-sdk) ✗ which python3
/opt/homebrew/bin/python3
➜  python-sdk git:(feat-python-sdk) ✗ python3 --version
Python 3.12.5
➜  python-sdk git:(feat-python-sdk) ✗ ls python/eppo_client
__init__.py                        _eppo_client.pyi                   config.py
__pycache__                        assignment_logger.py               py.typed
_eppo_client.cpython-311-darwin.so bandit.py                          version.py

➜  python-sdk git:(feat-python-sdk) ✗ pip install .
Processing /Users/leo/src/rust-sdk/python-sdk
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Building wheels for collected packages: eppo-server-sdk
  Building wheel for eppo-server-sdk (pyproject.toml) ... done
  Created wheel for eppo-server-sdk: filename=eppo_server_sdk-4.0.0-cp311-cp311-macosx_11_0_arm64.whl size=2235806 sha256=3983ad3721e3ff5c7d941f55e90670b8bada7b8bbb0de72ed825f7bf42d12bf5
  Stored in directory: /Users/leo/Library/Caches/pip/wheels/05/a2/8b/febb161210fb0561c324e64b7ff9400edef701dbe616d25850
Successfully built eppo-server-sdk
Installing collected packages: eppo-server-sdk
  Attempting uninstall: eppo-server-sdk
    Found existing installation: eppo-server-sdk 4.0.0
    Uninstalling eppo-server-sdk-4.0.0:
      Successfully uninstalled eppo-server-sdk-4.0.0
Successfully installed eppo-server-sdk-4.0.0

[notice] A new release of pip is available: 24.1.2 -> 24.2
[notice] To update, run: python3.11 -m pip install --upgrade pip
```

**Review by leoromanovsky — APPROVED** [review]
Ran it in a test python server and it operated as expected.

Really appreciate all the comments you added and explaining your rationale for decisions.

**Comment by rasendubi** [comment]
Going to merge this, so next PRs take python into account. Will coordinate with you to release it later
